File: cpp/tests/test_eof.cpp
#include "eof/eof.hpp"

#include <cmath>
#include <iostream>
#include <vector>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace {

int g_failed = 0;
int g_passed = 0;

void check(bool cond, const char* msg) {
    if (cond) {
        ++g_passed;
    } else {
        ++g_failed;
        std::cerr << "FAIL: " << msg << "\n";
    }
}

}  // namespace

int main() {
    using eof::Matrix;

    {
        const int n_loc = 10;
        const int n_time = 40;
        Matrix X(n_loc, n_time);
        for (int i = 0; i < n_loc; ++i) {
            for (int t = 0; t < n_time; ++t) {
                X(i, t) = std::sin(2.0 * M_PI * i / n_loc) *
                              std::cos(2.0 * M_PI * t / n_time) +
                          0.25 * std::cos(2.0 * M_PI * i / n_loc) *
                              std::sin(4.0 * M_PI * t / n_time);
            }
        }
        eof::PackedEof solver(X, 2);
        Matrix Xa = X;
        Xa.colwise() -= X.rowwise().mean();
        const double rel = (solver.reconstruct() - Xa).norm() / (Xa.norm() + 1e-15);
        check(rel < 1e-10, "rank-2 reconstruction");
        Matrix gram = solver.modes().transpose() * solver.modes();
        check((gram - Matrix::Identity(2, 2)).norm() < 1e-8, "orthonormal EOFs");
        check(solver.expvar()(0) > solver.expvar()(1), "ordered expvar");
    }

    {
        eof::PackedEof solver(Matrix::Random(6, 15));
        check(std::abs(solver.expvar().sum() - 100.0) < 1e-6, "expvar sums to 100");
    }

    {
        const int n_lon = 4, n_lat = 3, n_time = 50;
        std::vector<double> data(static_cast<size_t>(n_lon * n_lat * n_time), 0.0);
        const int nsp = n_lon * n_lat;
        for (int t = 0; t < n_time; ++t) {
            for (int j = 0; j < n_lat; ++j) {
                for (int i = 0; i < n_lon; ++i) {
                    const int s = i + j * n_lon;
                    double v = 0.0;
                    if (j == 1) {
                        v = std::sin(2.0 * M_PI * t / 20.0);
                    } else if (j == 2) {
                        v = std::sin(2.0 * M_PI * t / 13.0 + 0.4);
                    }
                    data[static_cast<size_t>(s + t * nsp)] = v;
                }
            }
        }
        for (int t = 0; t < n_time; ++t) {
            data[static_cast<size_t>(t * nsp)] = std::nan("");
        }
        double lat[3] = {-75.0, 0.0, 75.0};
        eof::GridOptions opt;
        opt.n_eof = 2;
        eof::GridEof weighted(data.data(), n_lon, n_lat, n_time, lat, 3, opt);
        eof::GridEof unweighted(data.data(), n_lon, n_lat, n_time, nullptr, 0, opt);
        auto maps = weighted.maps();
        check(std::isnan(maps[0]), "NaN mask preserved");
        check(std::isfinite(maps[1]), "ocean cell finite");
        const double ratio_w = weighted.expvar()(0) / weighted.expvar()(1);
        const double ratio_u = unweighted.expvar()(0) / unweighted.expvar()(1);
        check(ratio_w > ratio_u * 1.05, "area weight prefers equator");
    }

    {
        const int n_lon = 3, n_lat = 2, n_time = 40;
        const int nsp = n_lon * n_lat;
        std::vector<double> d1(static_cast<size_t>(nsp * n_time));
        std::vector<double> d2(static_cast<size_t>(nsp * n_time));
        for (int t = 0; t < n_time; ++t) {
            const double a = std::sin(2.0 * M_PI * t / 10.0);
            const double b = std::cos(2.0 * M_PI * t / 10.0);
            for (int p = 0; p < nsp; ++p) {
                d1[static_cast<size_t>(p + t * nsp)] = a;
                d2[static_cast<size_t>(p + t * nsp)] = 100.0 * b;
            }
        }
        eof::SpatialWeights ones(n_lon, n_lat);
        eof::MultivariateEof raw(d1.data(), n_lon, n_lat, d2.data(), n_lon, n_lat, n_time,
                                 ones, ones, false, 2);
        eof::MultivariateEof norm(d1.data(), n_lon, n_lat, d2.data(), n_lon, n_lat, n_time,
                                  ones, ones, true, 2);
        check(raw.expvar()(0) / raw.expvar()(1) > 10.0, "MEOF raw variance imbalance");
        check(std::abs(norm.expvar()(0) - norm.expvar()(1)) < 5.0,
              "MEOF normalize balances variance");
    }

    std::cout << "passed=" << g_passed << " failed=" << g_failed << "\n";
    return g_failed == 0 ? 0 : 1;
}

File: python/tests/test_eof.py
from __future__ import annotations

import numpy as np
import pytest

from eof_toolbox import Eof, MultivariateEof, PackedEof, SpatialWeights, eof, eofcore, latitude_weights, meof
from eof_toolbox.weights import make_weights


def _rank2_field(n_lon=8, n_lat=6, n_time=40):
    x = np.arange(n_lon)[:, None, None]
    y = np.arange(n_lat)[None, :, None]
    t = np.arange(n_time)[None, None, :]
    p1 = np.sin(2 * np.pi * x / n_lon) * np.cos(2 * np.pi * y / n_lat)
    p2 = np.cos(2 * np.pi * x / n_lon) * np.sin(2 * np.pi * y / n_lat)
    a1 = np.sin(2 * np.pi * t / n_time)
    a2 = np.cos(4 * np.pi * t / n_time)
    return p1 * a1 + 0.3 * p2 * a2


def test_core_orthonormal_and_reconstruction():
    rng = np.random.default_rng(0)
    X = rng.standard_normal((12, 30))
    L, Y, eig, expvar = eofcore(X, n_eof=5)
    gram = L.T @ L
    assert np.allclose(gram, np.eye(5), atol=1e-8)
    Xa = X - X.mean(axis=1, keepdims=True)
    reco = L @ Y
    # 5 modes cannot reconstruct all of a random matrix; residual must
    # still drop vs the mean-only baseline.
    assert np.linalg.norm(Xa - reco) < np.linalg.norm(Xa)
    assert expvar[0] >= expvar[-1]
    assert eig[0] >= eig[-1]


def test_full_rank_expvar_sums_to_100():
    rng = np.random.default_rng(1)
    data = rng.standard_normal((5, 4, 20))
    _, _, expvar, _ = eof(data)
    assert expvar.shape[0] == min(20, 20)
    assert abs(expvar.sum() - 100.0) < 1e-6


def test_rank2_grid_reconstruction():
    data = _rank2_field()
    maps, pcs, expvar, _ = eof(data, n_eof=2)
    anom = data - data.mean(axis=2, keepdims=True)
    reco = np.einsum("ijm,mt->ijt", maps, pcs)
    rel = np.linalg.norm(reco - anom) / np.linalg.norm(anom)
    assert rel < 1e-8
    assert expvar[0] > expvar[1]


def test_nan_mask():
    rng = np.random.default_rng(2)
    data = rng.standard_normal((5, 4, 12))
    data[1, 1, :] = np.nan
    maps, *_ = eof(data, n_eof=3)
    assert np.all(np.isnan(maps[1, 1, :]))
    assert np.all(np.isfinite(maps[0, 0, :]))


def test_latitude_weights_sqrtcos():
    lat = np.array([-60.0, 0.0, 60.0])
    W = latitude_weights(lat, n_lon=2, method="sqrtcos")
    assert W.shape == (2, 3)
    assert np.allclose(W[0], W[1])
    assert W[0, 1] > W[0, 0]
    assert np.allclose(W[0, 1], 1.0)


def test_area_weight_prefers_equator():
    lat = np.array([-75.0, 0.0, 75.0])
    n_lon, n_time = 4, 80
    t = np.arange(n_time)
    data = np.zeros((n_lon, 3, n_time))
    data[:, 1, :] = np.sin(2 * np.pi * t / 20)
    data[:, 2, :] = np.sin(2 * np.pi * t / 13 + 0.4)
    _, _, exp_unw, _ = eof(data, n_eof=2)
    _, _, exp_w, _ = eof(data, lat=lat, n_eof=2, weighting="sqrtcos")
    assert (exp_w[0] / exp_w[1]) > (exp_unw[0] / exp_unw[1]) * 1.05


def test_custom_weights_shape():
    data = _rank2_field(n_lon=4, n_lat=3, n_time=15)
    W = np.linspace(0.5, 1.5, 4 * 3).reshape(4, 3)
    maps_a, *_ = eof(data, weights=W, n_eof=2)
    maps_b, *_ = eof(data, weights=W.reshape(-1, order="F"), n_eof=2)
    assert np.allclose(maps_a, maps_b, equal_nan=True)


def test_standardize_scales_pcs():
    data = _rank2_field()
    _, pcs, _, eig = eof(data, n_eof=2, standardize=False)
    _, pcs_s, _, _ = eof(data, n_eof=2, standardize=True)
    ratio = np.std(pcs[0]) / np.std(pcs_s[0])
    assert np.isclose(ratio, np.sqrt(eig[0]), rtol=1e-6)


def test_meof_normalize_balances_variance():
    n_lon, n_lat, n_time = 4, 3, 40
    t = np.arange(n_time)
    a = np.sin(2 * np.pi * t / 10)
    b = np.cos(2 * np.pi * t / 10)
    d1 = np.broadcast_to(a, (n_lon, n_lat, n_time)).copy()
    d2 = 100.0 * np.broadcast_to(b, (n_lon, n_lat, n_time)).copy()
    *_, exp_raw, _ = meof(d1, d2, n_eof=2, normalize=False)
    *_, exp_norm, _ = meof(d1, d2, n_eof=2, normalize=True)
    assert exp_raw[0] / exp_raw[1] > 10
    assert abs(exp_norm[0] - exp_norm[1]) < 5.0


def test_meof_time_mismatch():
    with pytest.raises(ValueError):
        meof(np.zeros((2, 2, 5)), np.zeros((2, 2, 6)))


def test_make_weights_rejects_bad_lat():
    with pytest.raises(ValueError):
        make_weights(2, 2, lat=np.array([-10.0, 100.0]))


def test_eof_class_matches_function():
    data = _rank2_field()
    solver = Eof(data, n_eof=2)
    maps, pcs, expvar, eig = eof(data, n_eof=2)
    assert np.allclose(solver.maps(), maps, equal_nan=True)
    assert np.allclose(solver.pcs(), pcs)
    assert np.allclose(solver.expvar, expvar)
    reco = solver.reconstruct()
    anom = data - data.mean(axis=2, keepdims=True)
    assert np.linalg.norm(reco - anom) / np.linalg.norm(anom) < 1e-8


def test_packed_eof_class():
    rng = np.random.default_rng(3)
    X = rng.standard_normal((8, 20))
    solver = PackedEof(X, n_eof=3)
    L, Y, eig, expvar = eofcore(X, n_eof=3)
    assert np.allclose(solver.modes, L)
    assert np.allclose(solver.pcs, Y)
    assert np.allclose(solver.eigenvalues, eig)
    assert np.allclose(solver.expvar, expvar)


def test_spatial_weights_class():
    lat = np.array([-30.0, 0.0, 30.0])
    W = SpatialWeights.from_latitude(lat, n_lon=5)
    assert W.array.shape == (5, 3)
    assert np.allclose(W.array, latitude_weights(lat, 5))


def test_multivariate_eof_class():
    n_lon, n_lat, n_time = 4, 3, 40
    t = np.arange(n_time)
    d1 = np.broadcast_to(np.sin(2 * np.pi * t / 10), (n_lon, n_lat, n_time)).copy()
    d2 = 100.0 * np.broadcast_to(np.cos(2 * np.pi * t / 10), (n_lon, n_lat, n_time)).copy()
    solver = MultivariateEof(d1, d2, n_eof=2, normalize=True)
    m1, m2, pcs, expvar, eig = meof(d1, d2, n_eof=2, normalize=True)
    cm1, cm2 = solver.maps()
    assert np.allclose(cm1, m1, equal_nan=True)
    assert np.allclose(cm2, m2, equal_nan=True)
    assert np.allclose(solver.pcs(), pcs)
    assert np.allclose(solver.expvar, expvar)
    assert np.allclose(solver.eigenvalues, eig)
